programs/account-compression/src/instructions/nullify_leaves.rs

```diff
@@ -32,18 +34,40 @@ pub fn process_nullify_leaves<'a, 'b, 'c: 'info, 'info>(
     indices: &'a [u64],
     proofs: &'a [Vec<[u8; 32]>],
 ) -> Result<()> {
-    let mut array_account = ctx.accounts.indexed_array.load_mut()?;
-    if array_account.associated_merkle_tree != ctx.accounts.merkle_tree.key() {
-        msg!(
+    {
+        let array_account = ctx.accounts.indexed_array.load()?;
+        if array_account.associated_merkle_tree != ctx.accounts.merkle_tree.key() {
+            msg!(
             "Nullifier queue and Merkle tree are not associated. Associated mt of nullifier queue {} != merkle tree {}",
             array_account.associated_merkle_tree,
             ctx.accounts.merkle_tree.key(),
         );
-        return Err(AccountCompressionErrorCode::InvalidMerkleTree.into());
+            return Err(AccountCompressionErrorCode::InvalidMerkleTree.into());
+        }
+        drop(array_account);
     }
 
-    let leaf: [u8; 32] = array_account.indexed_array[leaves_queue_indices[0] as usize].element;
-    msg!("leaf {:?}", leaf);
+    let sequence_number = {
+        let mut merkle_tree_account = ctx.accounts.merkle_tree.load_mut()?;
+        let loaded_merkle_tree = merkle_tree_account.load_merkle_tree_mut()?;
+        loaded_merkle_tree.sequence_number
+    };
+
+    let leaf = {
+        let indexed_array = ctx.accounts.indexed_array.to_account_info();
+        let mut indexed_array = indexed_array.try_borrow_mut_data()?;
+        let mut indexed_array =
+            unsafe { indexed_array_from_bytes_zero_copy_mut(&mut indexed_array).unwrap() };
+        let leaf_cell = indexed_array
+            .by_value_index(leaves_queue_indices[0] as usize, None)
+            .cloned()
+            .ok_or(AccountCompressionErrorCode::LeafNotFound)?;
+        indexed_array
+            .mark_with_sequence_number(&leaf_cell.value_biguint(), sequence_number)
```

Comment:
shouldn't this be sequence number plus threshold?

Also I think this should happen after updating the Merkle tree otherwise your threshold is not high enough.

Reply:
The threshold is applied on the side of has set.

Reply:
To be precise - the threshold is applied here. And it was your suggestion, BTW. :smile:

Reply:
> Also I think this should happen after updating the Merkle tree otherwise your threshold is not high enough.
Reasoning about whether we should mark_with_sequence before or after the Merkle tree update.

Example with the current code ``mark_with_sequence`` before the Merkle tree updates: 
- root history array length is **2400**
- ``current_root_history_array_index`` = 10
- consider we have 2 nullifiers (``nullifier_0``, ``nullifier_1``) that we insert into the same Merkle tree
- ``mark_with_sequence`` marks both with ``10 + 2400``
- nullify ``nullifier_0`` creates a new root and increments ``current_root_history_array_index`` = 11
  - this root doesn't contain ``nullifier_0`` but still contains ``nuliffier_1``
  - however ``nullifier_1`` is marked with 10 not 11 -> there will be one root that you can still prove inclusion of and already overwrite ``nullifier_1``

Solution:
- move code that executes ``mark_with_sequence`` after update or add nullifiers.len() to sequence number

Do you follow?

Reply:
Good point, I think that marking with sequence after update makes sense.

Reply:
thanks!